File: Python/lark_card_kit/utils/element_finder.py
"""Element finder for locating and updating elements by ID or other criteria."""

from __future__ import annotations

from typing import Any, Callable, TypeVar

from ..models.elements import Element

E = TypeVar("E", bound=Element)


class ElementFinder:
    """Element finder for locating and updating elements within a card.

    Provides methods to find elements by ID, tag, or other criteria.
    Also supports updating elements by ID.

    示例:
        >>> from lark_card_kit import CardBuilder, ElementFinder
        >>>
        >>> # 创建卡片
        >>> with CardBuilder.create() as builder:
        ...     with builder.body() as b:
        ...         b.plain_text("Hello")
        ...         with b.button() as btn:
        ...             btn.text("Click").element_id("btn1")
        >>>
        >>> card = builder.build()
        >>>
        >>> # 查找并修改组件
        >>> finder = ElementFinder(card.body.elements)
        >>> button = finder.find_by_id("btn1")
        >>> if button:
        ...     button.disabled = True
        >>>
        >>> # 使用 update_by_id 便捷方法
        >>> finder.update_by_id("btn1", lambda e: setattr(e, "disabled", True))
    """

    def __init__(self, elements: list[Element]) -> None:
        """Initialize the element finder.

        Args:
            elements: List of elements to search.
        """
        self._elements = elements

    def find_by_id(self, element_id: str) -> Element | None:
        """Find an element by its ID.

        Args:
            element_id: Element ID to search for.

        Returns:
            Element if found, None otherwise.
        """
        return self._find_by_id_recursive(self._elements, element_id)

    def find_by_id_as(self, element_id: str, element_type: type[E]) -> E | None:
        """Find an element by its ID and cast to specific type.

        Args:
            element_id: Element ID to search for.
            element_type: Expected element type.

        Returns:
            Element if found and of correct type, None otherwise.
        """
        element = self.find_by_id(element_id)
        if isinstance(element, element_type):
            return element
        return None

    def update_by_id(self, element_id: str, updater: Callable[[Element], None]) -> bool:
        """Find an element by ID and update it.

        Args:
            element_id: Element ID to search for.
            updater: Function to update the element.

        Returns:
            True if element was found and updated, False otherwise.

        示例:
            >>> finder.update_by_id("btn1", lambda e: setattr(e, "disabled", True))
        """
        element = self.find_by_id(element_id)
        if element is not None:
            updater(element)
            return True
        return False

    def update_by_id_as(
        self,
        element_id: str,
        element_type: type[E],
        updater: Callable[[E], None],
    ) -> bool:
        """Find an element by ID, cast to specific type, and update it.

        Args:
            element_id: Element ID to search for.
            element_type: Expected element type.
            updater: Function to update the element.

        Returns:
            True if element was found, of correct type, and updated, False otherwise.

        示例:
            >>> from lark_card_kit import Button
            >>> finder.update_by_id_as("btn1", Button, lambda b: b.disabled = True)
        """
        element = self.find_by_id_as(element_id, element_type)
        if element is not None:
            updater(element)
            return True
        return False

    def _find_by_id_recursive(self, elements: list[Element], element_id: str) -> Element | None:
        """Recursively search for an element by ID.

        Args:
            elements: List of elements to search.
            element_id: Element ID to search for.

        Returns:
            Element if found, None otherwise.
        """
        for element in elements:
            if getattr(element, "element_id", None) == element_id:
                return element

            # Check nested elements
            nested = self._get_nested_elements(element)
            if nested:
                found = self._find_by_id_recursive(nested, element_id)
                if found:
                    return found

        return None

    def find_by_tag(self, tag: str) -> list[Element]:
        """Find all elements with a specific tag.

        Args:
            tag: Element tag to search for.

        Returns:
            List of matching elements.
        """
        return self._find_by_tag_recursive(self._elements, tag)

    def _find_by_tag_recursive(self, elements: list[Element], tag: str) -> list[Element]:
        """Recursively search for elements by tag.

        Args:
            elements: List of elements to search.
            tag: Element tag to search for.

        Returns:
            List of matching elements.
        """
        result: list[Element] = []

        for element in elements:
            if getattr(element, "tag", None) == tag:
                result.append(element)

            # Check nested elements
            nested = self._get_nested_elements(element)
            if nested:
                result.extend(self._find_by_tag_recursive(nested, tag))

        return result

    def update_all_by_tag(self, tag: str, updater: Callable[[Element], None]) -> int:
        """Find all elements with a specific tag and update them.

        Args:
            tag: Element tag to search for.
            updater: Function to update each element.

        Returns:
            Number of elements updated.
        """
        elements = self.find_by_tag(tag)
        for element in elements:
            updater(element)
        return len(elements)

    def _get_nested_elements(self, element: Element) -> list[Element]:
        """Get nested elements from a container element.

        Args:
            element: Container element.

        Returns:
            List of nested elements.
        """
        # Check for common container properties
        if hasattr(element, "elements") and isinstance(element.elements, list):
            return element.elements
        if hasattr(element, "columns") and isinstance(element.columns, list):
            result: list[Element] = []
            for column in element.columns:
                if hasattr(column, "elements"):
                    result.extend(column.elements)
            return result

        return []

File: TypeScript/LarkCardKit/src/utils/ElementFinder.ts
import type { Card } from '../models';
import type { Element } from '../models/elements';

/**
 * Element finder for locating and updating elements within a card.
 *
 * 提供通过 ID 或 tag 查找和修改组件的能力。
 *
 * @example
 * ```typescript
 * const finder = new ElementFinder(card);
 *
 * // 查找组件
 * const button = finder.findElementById('btn1');
 *
 * // 修改组件
 * finder.modifyElementById('btn1', (el) => {
 *   el.disabled = true;
 * });
 *
 * // 批量修改
 * finder.updateAllByTag('button', (el) => {
 *   el.disabled = true;
 * });
 * ```
 */
export class ElementFinder {
  constructor(private _card: Card) {}

  /**
   * 通过 ID 查找组件
   * @param elementId 组件 ID
   * @returns 找到的组件，如果未找到则返回 undefined
   */
  findElementById(elementId: string): Element | undefined;

  /**
   * 通过 ID 查找组件并验证类型
   * @param elementId 组件 ID
   * @param expectedTag 期望的组件 tag
   * @returns 找到的组件，如果未找到或类型不匹配则返回 undefined
   */
  findElementById<T extends Element>(elementId: string, expectedTag: string): T | undefined;

  findElementById(elementId: string, expectedTag?: string): Element | undefined {
    return this.findElementInElements(this._card.body.elements, elementId, expectedTag);
  }

  /**
   * 通过 ID 修改组件
   * @param elementId 组件 ID
   * @param modifier 修改函数
   * @returns 是否成功修改（组件存在时返回 true）
   */
  modifyElementById(elementId: string, modifier: (element: Element) => void): boolean {
    const element = this.findElementById(elementId);
    if (!element) return false;
    modifier(element);
    return true;
  }

  /**
   * 通过 ID 修改组件（带类型验证）
   * @param elementId 组件 ID
   * @param expectedTag 期望的组件 tag
   * @param modifier 修改函数
   * @returns 是否成功修改（组件存在且类型匹配时返回 true）
   */
  modifyElementByIdAs<T extends Element>(
    elementId: string,
    expectedTag: string,
    modifier: (element: T) => void
  ): boolean {
    const element = this.findElementById<T>(elementId, expectedTag);
    if (!element) return false;
    modifier(element);
    return true;
  }

  /**
   * 通过 tag 查找所有组件
   * @param tag 组件 tag
   * @returns 找到的组件列表
   */
  findAllElementsByTag(tag: string): Element[] {
    const results: Element[] = [];
    this.findElementsByTagInElements(this._card.body.elements, tag, results);
    return results;
  }

  /**
   * 批量修改所有指定 tag 的组件
   * @param tag 组件 tag
   * @param modifier 修改函数
   * @returns 修改的组件数量
   */
  updateAllByTag(tag: string, modifier: (element: Element) => void): number {
    const elements = this.findAllElementsByTag(tag);
    for (const element of elements) {
      modifier(element);
    }
    return elements.length;
  }

  private findElementInElements(elements: Element[], elementId: string, expectedTag?: string): Element | undefined {
    for (const element of elements) {
      if (element.element_id === elementId) {
        if (!expectedTag || element.tag === expectedTag) {
          return element;
        }
      }

      const elem = element as Record<string, unknown>;
      if ('elements' in elem && Array.isArray(elem.elements)) {
        const found = this.findElementInElements(elem.elements as Element[], elementId, expectedTag);
        if (found) return found;
      }

      if ('columns' in elem && Array.isArray(elem.columns)) {
        for (const column of elem.columns as Array<Record<string, unknown>>) {
          if (column.elements && Array.isArray(column.elements)) {
            const found = this.findElementInElements(column.elements as Element[], elementId, expectedTag);
            if (found) return found;
          }
        }
      }
    }
    return undefined;
  }

  private findElementsByTagInElements(elements: Element[], tag: string, results: Element[]): void {
    for (const element of elements) {
      if (element.tag === tag) {
        results.push(element);
      }

      const elem = element as Record<string, unknown>;
      if ('elements' in elem && Array.isArray(elem.elements)) {
        this.findElementsByTagInElements(elem.elements as Element[], tag, results);
      }

      if ('columns' in elem && Array.isArray(elem.columns)) {
        for (const column of elem.columns as Array<Record<string, unknown>>) {
          if (column.elements && Array.isArray(column.elements)) {
            this.findElementsByTagInElements(column.elements as Element[], tag, results);
          }
        }
      }
    }
  }
}